### DISCUSSION TITLE
Custom Endpoint For Custom Query or Modify Response?

### DISCUSSION THREAD

**Discussion by ghost** [pocketbase/pocketbase#5947] [Q&A] [answered]
Hi there, quick question about custom endpoints.

We got latitude / longitude coordinates in one of our collections. We need to send lat/lon/radio as parameters to endpoint and then query the collection using the Haversine distance to return only rows in the radio. 

If this is not the correct approach, will it work modifying the default API endpoint response?

I can't this to work for some reason. Here's our code:

Thank you!

```routerAdd("GET", "/custom/t_historico_mensual", async (c) => {
    try {
        // Parse the `tipo_operacion` parameter from the query string
        const tipo_operacion = c.queryParam("tipo_operacion");
        const latitud = parseFloat(c.queryParam("latitud"));
        const longitud = parseFloat(c.queryParam("longitud"));
        const radio = parseFloat(c.queryParam("radio"));

        // Validate the parameter
        if (!tipo_operacion) {
            throw new ApiError(400, "Invalid or missing 'tipo_operacion' parameter.");
        }

        // Define the structure of the expected results
        const result = arrayOf(new DynamicModel({
            id: "",
            aid: 0,
            latitud: "", // Set this to string for now since it shows an error about float/number types otherwise
            longitud: "", // Set this to string for now since it shows an error about float/number types otherwise
            barrio: "",
            valor: 0,
        }));

        // Execute the query
        await $app.dao().db()
    .select(
        "id",
        "aid",
        "CAST(latitud AS REAL) AS latitud",
        "CAST(longitud AS REAL) AS longitud",
        "barrio",
        "valor"
    )
    .from("t_historico_mensual")
    .andWhere($dbx.like("tipo_operacion", tipo_operacion)) // Match tipo_operacion
    .andWhere($dbx.like(`
        (6371 * acos(
            cos(radians(?)) * cos(radians("latitud")) *
            cos(radians("longitud") - radians(?)) +
            sin(radians(?)) * sin(radians("latitud"))
        )) <= (? / 1000)
    `, [latitud, longitud, latitud, radio]))
    .orderBy("created ASC") // Ensure 'created' exists in the table
    .limit(1000)
    .all(result);

        // Return the results
        return c.json(200, result);
    } catch (error) {
        console.error("Error in /custom/t_historico_mensual:", error);
        throw new ApiError(500, "Internal Server Error");
    }
});```

**Comment by ganigeorgiev** [comment, marked as answer]
> will it work modifying the default API endpoint response?

No. If you want to execute custom parameterized query you'll need a custom route.

I'm not familiar with the algorithm calculating the Haversine distance but the last dbx expression doesn't look correct, it should be:

```js
.andWhere($dbx.exp(`
    (6371 * acos(
        cos(radians({:latitud})) * cos(radians("latitud")) *
        cos(radians("longitud") - radians({:longitud})) +
        sin(radians({:latitud})) * sin(radians("latitud"))
    )) <= ({:radio} / 1000)
`, { latitud, longitud, radio}))
```

Also if you want an exact/equal match for the `tipo_operacion`, use `$dbx.hashExp({ tipo_operacion })` instead.

_Since the code sample have `$app.dao()` I assume that you use PocketBase <= v0.22. You can find the v0.22.x documentation at https://pocketbase.io/old/docs/js-database/._

**Reply by ghost** [reply]
Wow, I spend the whole weekend trying to figure this out. That worked, we can now query by radio.

Thank you!

How can we fix: GoError: sql: Scan error on column index 2, name "latitud": converting driver.Value type float64 ("-34.604269") to a int64: invalid syntax
```

// Define the structure of the expected results
        const result = arrayOf(new DynamicModel({
            id: "",
            aid: 0,
            latitud: 0.0,
            longitud: 0.0,
            barrio: "",
            valor: 0,
        }));
```
Our column in sqlite is NUMERIC, we don't understand how set it as REAL or if it is needed for this to work.

<img width="674" alt="image" src="https://github.com/user-attachments/assets/3a382ef0-e8b2-49b9-9d82-931f55a24ca1">

**Reply by ganigeorgiev** [reply]
This is similar to https://github.com/pocketbase/pocketbase/issues/5508 (it is mentioned alongside the new documentation examples).

To enforce float value type you can change the initial/zero value to `-0`, aka.:

```js
// Define the structure of the expected results
const result = arrayOf(new DynamicModel({
    id:       "",
    aid:      0,
    latitud:  -0,
    longitud: -0,
    barrio:   "",
    valor:    0,
}));
```

**Reply by ghost** [reply]
Thank Gani, that worked, I should I have searched better. Have a nice week!
